libr/core/disasm.c

```diff
@@ -3857,22 +3857,37 @@ static void ds_print_ptr(RDisasmState *ds, int len, int idx) {
 #endif
 }
 
+struct getreloc_t {
+	ut64 vaddr;
+	int size;
+};
+
+static int getreloc_tree(const void *user, const RBNode *n) {
+	struct getreloc_t *gr = (struct getreloc_t *)user;
+	const RBinReloc *r = container_of (n, const RBinReloc, vrb);
+	if ((r->vaddr >= gr->vaddr) && (r->vaddr < (gr->vaddr + gr->size))) {
+		return 0;
+	}
+
+	if (gr->vaddr > r->vaddr) {
+		return 1;
+	} else if (gr->vaddr < r->vaddr) {
+		return -1;
+	} else {
+		return 0;
+	}
+}
+
 // TODO: Use sdb in rbin to accelerate this
 // we shuold use aligned reloc addresses instead of iterating all of them
 static RBinReloc *getreloc(RCore *core, ut64 addr, int size) {
-	RList *list;
-	RBinReloc *r;
-	RListIter *iter;
 	if (size < 1 || addr == UT64_MAX) {
 		return NULL;
 	}
-	list = r_bin_get_relocs (core->bin);
-	r_list_foreach (list, iter, r) {
-		if ((r->vaddr >= addr) && (r->vaddr < (addr + size))) {
-			return r;
-		}
-	}
-	return NULL;
+	RBNode *relocs = r_bin_get_relocs (core->bin);
+	struct getreloc_t gr = { .vaddr = addr, .size = size };
+	RBNode *res = r_rbtree_find (relocs, &gr, getreloc_tree);
+	return res? container_of (res, RBinReloc, vrb): NULL;
```

Comment:
this container_of api is rbtree-specific, imho it should be named differently

Reply:
with r_rbtree_ prefix

Reply:
I don't think so. `container_of` is really rbtree specific. It's just a macro to get the address of an outer struct from the address of a field. It is used heavily in the linux kernel.

We should probably move it to r_util.h or in r_types.h instead of defining it in r_rbtree.h.